### ISSUE TITLE
Cls loss不正常

### ISSUE THREAD

**Issue by zhangbohnu** [zylo117/Yet-Another-EfficientDet-Pytorch#82]
您好！  非常感谢您的工作。
     我拉取了最新的代码，训练自己的数据集（只包含一个类别），obj_list已修改了。
训练时 Cls loss 很大。

tep: 40. Epoch: 0/10. Iteration: 41/54439. Cls loss: 3800.79980. Reg loss: 0.05531. Total loss: 3800.85522:   0%|

**Comment by zylo117** [comment]
just wait for a little bit longer, it will eventually drop. it's only 40 steps

**Comment by zhouhing** [comment]
我用自己的数据集转化为coco格式，和 voc格式，用这个最新的代码都跑了一遍，两个类别的class_loss仍然在0.77，mAP超级低，0.1不到，数据集在https://github.com/signatrix/efficientdet  上精度正常，我这个新手还是找不出原因，哈哈哈。然后用作者的efficientdet-d0.pth权重检测coco数据集中的图片，检测结果很奇怪
![image](https://user-images.githubusercontent.com/51043773/79298224-17493000-7f13-11ea-87c3-ff032fef03d8.png)
数据集之外的图片，也很奇怪，
![image](https://user-images.githubusercontent.com/51043773/79298402-9e96a380-7f13-11ea-89e6-60e2305f3d56.png)
![image](https://user-images.githubusercontent.com/51043773/79298479-d56cb980-7f13-11ea-8412-8fdcd80006dd.png)

可能是我这边数据集不适合这个模型的问题，不过还是很感谢作者的工作！

**Comment by zylo117** [comment]
are these picture from coco?
can you share their image filename?

**Comment by mad-fogs** [comment]
> 我用自己的数据集转化为coco格式，和 voc格式，用这个最新的代码都跑了一遍，两个类别的class_loss仍然在0.77，mAP超级低，0.1不到，数据集在https://github.com/signatrix/efficientdet 上精度正常，我这个新手还是找不出原因，哈哈哈。然后用作者的efficientdet-d0.pth权重检测coco数据集中的图片，检测结果很奇怪
> ![image](https://user-images.githubusercontent.com/51043773/79298224-17493000-7f13-11ea-87c3-ff032fef03d8.png)
> 数据集之外的图片，也很奇怪，
> ![image](https://user-images.githubusercontent.com/51043773/79298402-9e96a380-7f13-11ea-89e6-60e2305f3d56.png)
> ![image](https://user-images.githubusercontent.com/51043773/79298479-d56cb980-7f13-11ea-8412-8fdcd80006dd.png)
> 
> 可能是我这边数据集不适合这个模型的问题，不过还是很感谢作者的工作！

i have got similar result, change thresholds from test script will improve but still far from target performance.

**Comment by zhouhing** [comment]
> 这些照片来自可可吗？
> 可以分享他们的图片文件名吗？

第一张熊是coco里面的coco/val2017/000000000285.jpg
后面两张是网上的图

**Comment by ChrisHaooooooo** [comment]
I got a similar result from d0 like:
![image](https://user-images.githubusercontent.com/42533211/79302124-02be6500-7f1e-11ea-99c6-e480e6e740df.png)
while I use the yolov3-spp and got:
![image](https://user-images.githubusercontent.com/42533211/79302180-2da8b900-7f1e-11ea-91c4-563ea980ae45.png)

**Comment by zhangbohnu** [comment]
> just wait for a little bit longer, it will eventually drop. it's only 40 steps
ok,  thank you.

**Comment by mad-fogs** [comment]
> > 我用自己的数据集转化为coco格式，和 voc格式，用这个最新的代码都跑了一遍，两个类别的class_loss仍然在0.77，mAP超级低，0.1不到，数据集在https://github.com/signatrix/efficientdet 上精度正常，我这个新手还是找不出原因，哈哈哈。然后用作者的efficientdet-d0.pth权重检测coco数据集中的图片，检测结果很奇怪
> > ![image](https://user-images.githubusercontent.com/51043773/79298224-17493000-7f13-11ea-87c3-ff032fef03d8.png)
> > 数据集之外的图片，也很奇怪，
> > ![image](https://user-images.githubusercontent.com/51043773/79298402-9e96a380-7f13-11ea-89e6-60e2305f3d56.png)
> > ![image](https://user-images.githubusercontent.com/51043773/79298479-d56cb980-7f13-11ea-8412-8fdcd80006dd.png)
> > 可能是我这边数据集不适合这个模型的问题，不过还是很感谢作者的工作！
> 
> i have got similar result, change thresholds from test script will improve but still far from target performance.

@zylo117  i just changed the 2 default threshold and input_size value, the result comes to be normal.
wow, finally it works.

**Comment by zylo117** [comment]
> I got a similar result from d0 like:
> ![image](https://user-images.githubusercontent.com/42533211/79302124-02be6500-7f1e-11ea-99c6-e480e6e740df.png)
> while I use the yolov3-spp and got:
> ![image](https://user-images.githubusercontent.com/42533211/79302180-2da8b900-7f1e-11ea-91c4-563ea980ae45.png)

you can eliminate the overlapping boxes by reducing iou_threshold and increasing confidence_threshold

**Comment by ChrisHaooooooo** [comment]
@zylo117 Thanks! I set `threshold = 0.5` and `force_input_size = None` . After that I got a normal detection which is similar to the yolo's. But I dont understand why `force_input_size = None` is so important when I use d0? If I use 1920 , the bbox can't even accurately predict a person.

**Comment by zylo117** [comment]
> @zylo117 Thanks! I set `threshold = 0.5` and `force_input_size = None` . After that I got a normal detection which is similar to the yolo's. But I dont understand why `force_input_size = None` is so important when I use d0? If I use 1920 , the bbox can't even accurately predict a person.

1920 is for demo, because the objects in demo image is very small. You need to set `force_input_size = None` to get the best performance.

**Comment by Liu0329** [comment]
Same problem. Do you load pretrained weight ?

**Comment by YuaCC** [comment]
'''
如果是从头训练需要改一下初始化就可以了
utils/utils.py
'''
def init_weights(model):
    import math
    for name, module in model.named_modules():
        is_conv_layer = isinstance(module, nn.Conv2d)

        if is_conv_layer:
            nn.init.kaiming_uniform_(module.weight.data)

            if module.bias is not None:
                module.bias.data.zero_()
    prior = 0.01
    model.classifier.header.pointwise_conv.conv.weight.data.fill_(0)
    model.classifier.header.pointwise_conv.conv.bias.data.fill_(-math.log((1.0 - prior) / prior))

**Comment by YuaCC** [comment]
`'''
如果是从头训练需要改一下初始化就可以了
utils/utils.py
'''
def init_weights(model):
import math
for name, module in model.named_modules():
is_conv_layer = isinstance(module, nn.Conv2d)

    if is_conv_layer:
        nn.init.kaiming_uniform_(module.weight.data)

        if module.bias is not None:
            module.bias.data.zero_()
prior = 0.01
model.classifier.header.pointwise_conv.conv.weight.data.fill_(0)
model.classifier.header.pointwise_conv.conv.bias.data.fill_(-math.log((1.0 - prior) / prio`

**Comment by YuaCC** [comment]
`
def init_weights(model):
    import math
    for name, module in model.named_modules():
        is_conv_layer = isinstance(module, nn.Conv2d)

        if is_conv_layer:
            nn.init.kaiming_uniform_(module.weight.data)

            if module.bias is not None:
                module.bias.data.zero_()
    prior = 0.01
    model.classifier.header.pointwise_conv.conv.weight.data.fill_(0)
    model.classifier.header.pointwise_conv.conv.bias.data.fill_(-math.log((1.0 - prior) / prior))
`

**Comment by zylo117** [comment]
> Adding these three lines of code at the end of the function utils/utils.py/init_weights(model) can solve the problem@zylo117
> prior = 0.01
> 
> model.classifier.header.pointwise_conv.conv.weight.data.fill_(0)
> 
> model.classifier.header.pointwise_conv.conv.bias.data.fill_(-math.log((1.0 - prior) / prior))

Yes, it's fixed, the loss should be lower in the early stage now.

https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/blob/master/utils/utils.py#L224
